File: data/scripts/nwck.py
#http://rosalind.info/problems/nwck/

from collections import defaultdict, deque


def parse_newick(s):
    """Costruisce un albero da una stringa in formato Newick (senza lunghezze di ramo).
    Restituisce la lista di adiacenza (non orientata) e un dizionario nome -> id nodo."""
    s = s.strip()
    s = s.removesuffix(";")
    adiacenza = defaultdict(list)
    nomi = {}
    pos = [0]
    contatore = [0]

    def nuovo_nodo():
        contatore[0] += 1
        return contatore[0]

    def parse_clade():
        nodo = nuovo_nodo()
        if pos[0] < len(s) and s[pos[0]] == "(":
            pos[0] += 1
            while True:
                figlio = parse_clade()
                adiacenza[nodo].append(figlio)
                adiacenza[figlio].append(nodo)
                if s[pos[0]] == ",":
                    pos[0] += 1
                    continue
                elif s[pos[0]] == ")":
                    pos[0] += 1
                    break
        inizio_nome = pos[0]
        while pos[0] < len(s) and s[pos[0]] not in ",()":
            pos[0] += 1
        nome = s[inizio_nome:pos[0]]
        if nome:
            nomi[nome] = nodo
        return nodo

    parse_clade()
    return adiacenza, nomi

def distanza(adiacenza, nomi, x, y):
    partenza = nomi[x]
    arrivo = nomi[y]
    visitati = {partenza: 0}
    coda = deque([partenza])
    while coda:
        corrente = coda.popleft()
        if corrente == arrivo:
            return visitati[corrente]
        for vicino in adiacenza[corrente]:
            if vicino not in visitati:
                visitati[vicino] = visitati[corrente] + 1
                coda.append(vicino)
    return None

def lettura(filename):
    with open(filename) as f:
        righe = [r.strip() for r in f if r.strip() != ""]
    blocchi = []
    for i in range(0, len(righe), 2):
        albero = righe[i]
        x, y = righe[i + 1].split()
        blocchi.append((albero, x, y))
    return blocchi

def main():
    blocchi = lettura("dataset.txt")
    risultati = []
    for albero, x, y in blocchi:
        adiacenza, nomi = parse_newick(albero)
        risultati.append(str(distanza(adiacenza, nomi, x, y)))
    print(" ".join(risultati))

if __name__ == "__main__":
    # execute only if run as a script
    main()

File: public/js-solutions/nwck.mjs
// Distances in Trees (Rosalind ID: NWCK) - soluzione JavaScript
// indipendente, non una trascrizione di problem.py: stessa logica
// (parser ricorsivo discendente per il formato Newick senza lunghezze
// di ramo, costruzione della lista di adiacenza, poi BFS per la
// distanza tra due foglie nominate), riscritta in modo idiomatico per
// JS.
//
// Contratto: riceve il contenuto testuale di dataset.txt, restituisce
// l'output testuale (stessa forma dell'output Python).
function parseNewick(sInput) {
  let s = sInput.trim();
  if (s.endsWith(";")) s = s.slice(0, -1);

  const adiacenza = new Map();
  const nomi = new Map();
  const pos = { v: 0 };
  const contatore = { v: 0 };

  function getAdiacenza(nodo) {
    if (!adiacenza.has(nodo)) adiacenza.set(nodo, []);
    return adiacenza.get(nodo);
  }

  function nuovoNodo() {
    contatore.v += 1;
    return contatore.v;
  }

  function parseClade() {
    const nodo = nuovoNodo();
    if (pos.v < s.length && s[pos.v] === "(") {
      pos.v += 1;
      while (true) {
        const figlio = parseClade();
        getAdiacenza(nodo).push(figlio);
        getAdiacenza(figlio).push(nodo);
        if (s[pos.v] === ",") {
          pos.v += 1;
          continue;
        } else if (s[pos.v] === ")") {
          pos.v += 1;
          break;
        }
      }
    }
    const inizioNome = pos.v;
    while (pos.v < s.length && !",()".includes(s[pos.v])) {
      pos.v += 1;
    }
    const nome = s.slice(inizioNome, pos.v);
    if (nome) nomi.set(nome, nodo);
    return nodo;
  }

  parseClade();
  return { adiacenza, nomi };
}

function distanza(adiacenza, nomi, x, y) {
  const partenza = nomi.get(x);
  const arrivo = nomi.get(y);
  const visitati = new Map([[partenza, 0]]);
  const coda = [partenza];
  let head = 0;
  while (head < coda.length) {
    const corrente = coda[head++];
    if (corrente === arrivo) return visitati.get(corrente);
    for (const vicino of adiacenza.get(corrente) ?? []) {
      if (!visitati.has(vicino)) {
        visitati.set(vicino, visitati.get(corrente) + 1);
        coda.push(vicino);
      }
    }
  }
  return null;
}

function lettura(datasetText) {
  const righe = datasetText
    .split("\n")
    .map((r) => r.trim())
    .filter((r) => r !== "");

  const blocchi = [];
  for (let i = 0; i < righe.length; i += 2) {
    const albero = righe[i];
    const [x, y] = righe[i + 1].split(/\s+/);
    blocchi.push([albero, x, y]);
  }
  return blocchi;
}

export default function solve(datasetText) {
  const blocchi = lettura(datasetText);

  if (blocchi.length === 0) {
    throw new Error("Input non valido: nessun blocco albero/coppia trovato");
  }

  const risultati = [];
  for (const [albero, x, y] of blocchi) {
    const { adiacenza, nomi } = parseNewick(albero);
    risultati.push(String(distanza(adiacenza, nomi, x, y)));
  }

  return `${risultati.join(" ")}\n`;
}
